org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/edit/BatchModification.java

```diff
@@ -0,0 +1,178 @@
+			super();
+			this.object = object;
+			this.modification = modification;
+		}
+	}
+
+	private final static Logger LOG = Logger.getLogger(BatchModification.class);
+	@Inject
+	private ChangeConverter.Factory changeConverterFactory;
+	private IXtextDocument document;
+	@Inject
+	private LtkIssueAcceptor issueAcceptor;
+	@Inject
+	private LiveScopeResourceSetInitializer liveScopeResourceSetInitializer;
+	private IProject project;
+	@Inject
+	private IResourceSetProvider resourceSetProvider;
+	@Inject
+	private Provider<IChangeSerializer> serializerProvider;
+
+	public void apply(Iterable<IBatchableModification> modifications, IProgressMonitor monitor) {
+		if (Iterables.isEmpty(modifications)) {
+			return;
+		}
+		try {
+			new WorkspaceModifyOperation() {
+				@Override
+				protected void execute(IProgressMonitor monitor) throws CoreException, InvocationTargetException, InterruptedException {
+					applyInWorkspace(modifications, monitor);
+				}
+			}.run(monitor);
+		} catch (InvocationTargetException | InterruptedException e) {
+			LOG.error(e.getMessage(), e);
+		}
+	}
+
+	protected void applyInWorkspace(Iterable<IBatchableModification> modifications, IProgressMonitor monitor) throws CoreException {
+		IProject proj = getResolvedProject();
+		if (proj == null) {
+			LOG.error("No project configured.");
+			return;
+		}
+		ResourceSet resourceSet = resourceSetProvider.get(proj);
+		liveScopeResourceSetInitializer.initialize(resourceSet);
+		List<ResolvedModification> resolved = Lists.newArrayList();
+		for (IBatchableModification modification : modifications) {
+			EObject obj = resourceSet.getEObject(modification.getEObjectURI(), true);
+			if (obj == null || obj.eIsProxy()) {
+				LOG.error("Invlid EObject URI " + modification.getEObjectURI());
+				continue;
+			}
+			resolved.add(new ResolvedModification(obj, modification));
+		}
+		IChangeSerializer serializer = serializerProvider.get();
+		for (ResolvedModification mod : resolved) {
+			if (monitor.isCanceled()) {
+				throw new OperationCanceledException();
+			}
+			mod.modification.apply(mod.object, serializer);
+		}
+		boolean first = true;
+		for (ResolvedModification mod : resolved) {
+			if (first) {
+				serializer.setUpdateCrossReferences(mod.modification.isUpdateCrossReferences());
+				serializer.setUpdateRelatedFiles(mod.modification.isUpdateRelatedFiles());
+			} else {
+				if (serializer.isUpdateCrossReferences() != mod.modification.isUpdateCrossReferences()) {
+					LOG.error("two modifications can't be batched.");
+					return;
+				}
+				if (serializer.isUpdateRelatedFiles() != mod.modification.isUpdateRelatedFiles()) {
+					LOG.error("two modifications can't be batched.");
+					return;
+				}
+			}
+		}
+		if (monitor.isCanceled()) {
+			throw new OperationCanceledException();
+		}
+		ChangeConverter converter = changeConverterFactory.create("Resolving Issues", null, issueAcceptor);
+		serializer.applyModifications(converter);
```

Comment:
Should the serializer accept a cancel indicator or progress monitor? Or is this considered to be a fast operation?

Reply:
fixed.